File: computer4/2_sep_19/SOLUTIONS/lvl4_no_time_to_paint.java
import java.util.*;
public class Main {

	public static void main(String[] args) {
		Scanner k =new Scanner(System.in);
		
		int n = k.nextInt();
		int q = k.nextInt();
		
		String a = k.next();
		
		int[] pre = new int[n+1];
		int[] suf = new int[n+2];
		proc(a, pre);
		proc(new StringBuilder(a).reverse().toString(), suf);
		
//		System.out.println(Arrays.toString(pre));
//		System.out.println(Arrays.toString(suf));
		for(int l=1, r=n; l<r; l++, r--) {
			int tmp = suf[l]; suf[l]=suf[r]; suf[r] = tmp;
		}
		for(int i = 0; i < q; i++) {
			int s = k.nextInt();
			int e = k.nextInt();
			
			System.out.println(pre[s-1]+suf[e+1]);
		}
	}
	static void proc(String a, int pre[]) {
		boolean [] vis = new boolean[26];
		for(int i = 0; i < a.length(); i++) {
			int c = a.charAt(i) - 'A';  
			if(!vis[c]) pre[i+1] = pre[i] + 1;
			else pre[i+1] = pre[i];
			vis[c] = true;
			for(int j=c+1; j<26; j++) vis[j] = false;
		}
	}

}

File: computer4/2_sep_19/SOLUTIONS/lvl4_no_time_to_paint.cpp
#include <bits/stdc++.h>
using namespace std;
const int MM = 1e5+3;
int N, Q, pre[MM], suf[MM]; string s; bool vis[26];
void fun(string &s, int cnt[]){
    memset(vis, 0, sizeof(vis));
    for(int i=1; i<=N; i++){
        int c = s[i-1] - 'A';
        if(vis[c]) {
            cnt[i] = cnt[i-1];
        }else{
            cnt[i] = cnt[i-1] + 1;
        }
        vis[c] = true;
        for(int j=c+1; j<26; j++) vis[j] = false;
    }
}
int main(){
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> N >> Q >> s;
    fun(s, pre);
    reverse(s.begin(), s.end());
    fun(s, suf);
    reverse(suf+1, suf+N+1);
    for(int i=1, x, y; i<=Q; i++){
        cin >> x >> y;
        cout << pre[x-1] + suf[y+1] << "\n";
    }
}